### PR TITLE
Exclude Objects (Superscedes #4716)

### PR DESCRIPTION
**Pull request by kageurufu**
module: Exclude Objects

Adding Klipper functionality to support cancelling objects while printing.

This module keeps track of motion in and out of objects and adjusts movements as needed. It also
tracks object status and provides that to clients.

The Klipper module is relatively simple, and only provides one piece of the workflow. Support from Moonraker is underway to pre-process gcode files from various slicers with the extended gcode commands supplied by this module. UI's such as Fluidd and Mainsail will provide the user facing controls.

There has been a small group sharing code. In addition to the Moonraker work, I have Fluidd code that will be submitted shortly. Mainsail support is also underway.

Signed-off-by: Troy Jacobson [troy.d.jacobson@gmail.com](mailto:troy.d.jacobson@gmail.com)
Co-authored-by: Franklyn [voron@tackitt.net](mailto:voron@tackitt.net)
Co-authored-by: Eric Callahan [arksine.code@gmail.com](mailto:arksine.code@gmail.com)

This PR superscedes #4716, has the requested changes from the prior PR, and fixes the failing test. Troy is currently busy, so I'm opening a new PR to push things along

### CODE REVIEW

**Comment by nightime7** [comment]
I've been running this for a couple of weeks now. seems to work quite well. is there any update regarding delivering this to the master repo? @Arksine  @kageurufu

**Comment by gurusonwheels** [comment]
whats the status of this project ?

**Comment by mjoconr** [comment]
Has worked really well for me also. Be nice to know when it might be included in to Klipper.

**Comment by lbibass** [comment]
This has also been working very well for me. I would also like to see it merged soon so I can get back to mainline.

**Comment by JRHeaton** [comment]
Another +1 from me for seemingly stable working. Would love to see a review on this from @KevinOConnor and get it into main! This is one of those killer features for big bed batch printers.

**Comment by batista16** [comment]
+1 been using it for months now.. let's get it merged? :)

**Comment by gurusonwheels** [comment]
+1 please merge soon . this is needed like yesterday .

**Comment by obamin** [comment]
+1 please merge soon .

File: klippy/extras/exclude_object.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+                self._normal_move(newpos, speed)
+
+    cmd_EXCLUDE_OBJECT_START_help = "Marks the beginning the current object" \
+                                    " as labeled"
+    def cmd_EXCLUDE_OBJECT_START(self, gcmd):
+        name = gcmd.get('NAME').upper()
+        if not any(obj["name"] == name for obj in self.objects):
+            self._add_object_definition({"name": name})
+        self.current_object = name
+        self.was_excluded_at_start = self._test_in_excluded_region()
```
Comment (Arksine) [inline klippy/extras/exclude_object.py:202]: Stale attribute?  It doesn't appear that `was_excluded_at_start` is accessed in the module.

File: klippy/extras/exclude_object.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+    cmd_EXCLUDE_OBJECT_END_help = "Marks the end the current object"
+    def cmd_EXCLUDE_OBJECT_END(self, gcmd):
+        if self.current_object == None and self.next_transform:
+            gcmd.respond_info("EXCLUDE_OBJECT_END called, but no object is"
+                              " currently active")
+            return
+        name = gcmd.get('NAME', default=None)
+        if name != None and name.upper() != self.current_object:
```
Comment (Arksine) [inline klippy/extras/exclude_object.py:211]: Generally identity should be used when testing for a `NoneType`, ie:
```python
if self.current_object is None and self.next_transform:
    gcmd.respond_info("EXCLUDE_OBJECT_END called, but no object is"
                                     " currently active")
    return
if name is not None and name.upper() != self.current_object:
     # rest of code
```

File: klippy/extras/exclude_object.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+            center = parameters.pop('CENTER', None)
+            polygon = parameters.pop('POLYGON', None)
+
+            obj = {"name": name.upper()}
+            obj.update(parameters)
+
+            if center != None:
+                obj['center'] = json.loads('[%s]' % center)
+
+            if polygon != None:
```
Comment (Arksine) [inline klippy/extras/exclude_object.py:265]: The `center` and `polygon` conditions should test for identity.

**Review by Arksine — COMMENTED** [review]
I found some minor issues that I do not believe would functionally impair the module, overall the implementation looks good to me.  We could correct these after the merging, however I will leave that up to Kevin.  I would be willing to submit a PR to do so at his request.

**Comment by KevinOConnor** [comment]
Thanks.  Looks good to me.  I agree minor changes should just be made after merging.  It's been a few weeks since this was submitted so I'd like confirmation from @kageurufu that everything is still good for committing and then I will commit.

-Kevin

**Comment by kageurufu** [comment]
@KevinOConnor It's still working great on my machine, and a few others I know. I think we're good!

**Comment by KevinOConnor** [comment]
Thanks!

-Kevin

**Comment by lbibass** [comment]
Woohoo!!

**Comment by gurusonwheels** [comment]
think its missing something   console shows 
14:51:03
Unknown command:"DEFINE_OBJECT"
and the icon that normally there next to stop and pause for selecting objects is missing .

**Comment by Arksine** [comment]
The gcode commands changed.  You will need to make sure you are using a file processed by `preprocess-cancellation` version 0.2.0.

**Comment by gurusonwheels** [comment]
> The gcode commands changed. You will need to make sure you are using a file processed by `preprocess-cancellation` version 0.2.0.

thanks . got you something . check discord .

**Comment by Lenbok** [comment]
Just curious, what is the use case for this feature? Is it that the user wants to slice a whole bunch of items and just select at print time which ones to actually print? Is it that midway through printing a plate of parts, some parts fail, so this lets you cancel the failed part (preventing collateral damage and filament waste)? (It would be nice if the documentation gave hints when you would want to do this, as it potentially allows new workflows that people aren't aware of.)

**Comment by clee** [comment]
> Is it that midway through printing a plate of parts, some parts fail, so this lets you cancel the failed part (preventing collateral damage and filament waste)?

This is the primary use case as far as I'm concerned.

**Pull request merged**